Add complete systemd support

ecksun opened the pull request:
Now both alertmanager and prometheus have support for systemd.

This patch also contains a breaking change where it removes the flag
prometheus_node_exporter_use_systemd in favour of
prometheus_use_systemd, that configures all three services.

This solves issue #30

fjaeckel commented:
I'd love to see this merged!

mbolek commented:
@William-Yeh could you share some ETA? This one looks like a no-brainer :)

pieterlexis commented:
Why not use the `ansible_service_mgr` fact to determine systemd unit files should be installed?

ecksun commented:
> Why not use the `ansible_service_mgr` fact to determine systemd unit files should be installed?

Well, I didn't know that was a thing and wanted to keep the style of the playbook.

It looks like an excellent idea, however I think it would be more suitable to its own PR.

dhoppe commented on templates/alertmanager.service.j2:
```diff
@@ -0,0 +1,17 @@
+After=network.target
+
+[Service]
+WorkingDirectory={{ prometheus_alertmanager_db_path }}
```
Why do not you use `-storage.path={{ prometheus_alertmanager_db_path }}` instead?

ecksun replied:
That is a very good question, I have change dit to use -storage.path instead

cybergrind commented:
@William-Yeh, is that or #24 going to be merged? 
It looks like a quite important feature for everyone with systemd but repo has only 1 commit since last year, should it be considered as dead?

keimond commented:
daemon-reload is definitely needed.. on a fresh install of CentOS 7, when the playbook tries to start prometheus, it fails. So I go to the system.. "systemctl daemon-reload".. run the playbook again, and the service starts.

lhoss commented:
+1
ps: related, because of improving node-exporter restart handling: https://github.com/William-Yeh/ansible-prometheus/pull/38/files

ps: In the hope to bring the master branch up to speed: https://github.com/William-Yeh/ansible-prometheus/issues/64

czerasz commented on tasks/install-alertmanager.yml:
```diff
@@ -109,8 +109,16 @@
   copy: src="../files/etc-default-alertmanager"  dest=/etc/default/alertmanager
 
+- name: copy systemd config to server
+  template: src="../templates/alertmanager.service.j2"  dest="/lib/systemd/system/alertmanager.service"
```
Why `/lib/systemd/system/alertmanager.service` instead of `/etc/systemd/system/alertmanager.service`? I'm coming from [here](https://unix.stackexchange.com/a/208352/30457).

czerasz commented on tasks/install-prometheus.yml:
```diff
@@ -107,8 +107,16 @@
   copy: src="../files/etc-default-prometheus"  dest=/etc/default/prometheus
 
+- name: copy systemd config to server
+  template: src="../templates/prometheus.service.j2"  dest="/lib/systemd/system/prometheus.service"
```
Why `/lib/systemd/system/prometheus.service` instead of `/etc/systemd/system/prometheus.service`? I'm coming from [here](https://unix.stackexchange.com/a/208352/30457).

czerasz commented on handlers/main.yml:
```diff
@@ -2,6 +2,16 @@
 # file: handlers/main.yml
 #
 
+- name: reload systemd prometheus
```
My suggestion would be to remove `reload systemd prometheus`, `reload systemd alertmanager` and use this instead:

```yaml
- name: reload systemd
  systemd:
    daemon_reload: yes
```

ecksun replied:
I actually don't think you can do `daemon_reload` without a service

czerasz commented on tasks/install-node-exporter.yml:
```diff
@@ -105,11 +105,11 @@
 - name: copy systemd config to server
   template: src="../templates/node_exporter.service.j2"  dest="/lib/systemd/system/node_exporter.service"
-  when: prometheus_node_exporter_use_systemd|bool
+  when: prometheus_use_systemd|bool
```
Missing:

```yaml
notify:
  - reload systemd
  - restart node_exporter
```

It was missing for a longer time :smile:

czerasz commented on tasks/install-node-exporter.yml:
```diff
@@ -105,11 +105,11 @@
+  when: not prometheus_use_systemd|bool
 
 - name: copy systemd config to server
   template: src="../templates/node_exporter.service.j2"  dest="/lib/systemd/system/node_exporter.service"
```
Now I see why the other configuration files went to `/lib/systemd/system/`. But shouldn't this be `/etc/systemd/system/`...? I'm coming from [here](https://unix.stackexchange.com/a/208352/30457).

ecksun replied:
yes, you are correct

czerasz commented:
The change proposed in [PR 24](https://github.com/William-Yeh/ansible-prometheus/pull/24/files) should be added here as well.

`ansible-prometheus/templates/node_exporter.service.j2` should have at the end:

```
[Install]
WantedBy=multi-user.target
```

czerasz commented on templates/alertmanager.service.j2:
```diff
@@ -0,0 +1,17 @@
+User={{ prometheus_user }}
+Group={{ prometheus_group }}
+Restart=on-failure
+
```
Why not add
```
ExecReload=/bin/kill -SIGHUP $MAINPID
```

here as well?

ecksun commented:
@czerasz I incorporated your changes, but we have moved on to [cloudalchemy.alertmanager][0] so, as with this update, future updates might take some time :)

[0]: https://galaxy.ansible.com/cloudalchemy/alertmanager/

lae commented:
I too have moved on to the cloudalchemy roles for managing my Prometheus environments - given how heavy the issue/PR backlog has become and lack of movement from maintainers of this role.

anisse commented:
Same here. @William-Yeh what do you think of encouraging people to move to the maintained @cloudalchemy roles ?

paulfantom commented:
Come to @cloudalchemy, we have cookies :smile:

William-Yeh commented:
Updated since https://github.com/William-Yeh/ansible-prometheus/tree/2.5.0

Thank you.
